control/controlhttp/client.go

```diff
@@ -189,26 +188,34 @@ func (a *dialParams) dialURL(ctx context.Context, u *url.URL) (*controlbase.Conn
 // tryURLUpgrade connects to u, and tries to upgrade it to a net.Conn.
 //
 // Only the provided ctx is used, not a.ctx.
-func (a *dialParams) tryURLUpgrade(ctx context.Context, u *url.URL, init []byte) (net.Conn, error) {
+func (a *Dialer) tryURLUpgrade(ctx context.Context, u *url.URL, init []byte) (net.Conn, error) {
 	dns := &dnscache.Resolver{
 		Forward:          dnscache.Get().Forward,
 		LookupIPFallback: dnsfallback.Lookup,
 		UseLastGood:      true,
 	}
+
+	var dialer dnscache.DialContextFunc
+	if a.Dialer != nil {
+		dialer = a.Dialer
+	} else {
+		dialer = new(net.Dialer).DialContext
+	}
```

Comment:
you could write the strDefault as:

```go
func valOr[T comparable](v1, v2 T) T { ... }
```

And then use it here too.

But drop the `new(net.Dialer)` and make a package global `var stdDialer net.Dialer` and then just use `stdDialer.DialContext` here in the `valOr` call so it doesn't allocate?

Reply:
`dnscache.DialContextFunc does not implement comparable` unfortunately :unamused:

Reply:
well, could use T any and reflect.ValueOf(v1).IsZero(). I wish there was a magic constraint that's like `isZeroCheckable`